ocaml/xapi/repository.ml

```diff
@@ -1024,3 +1025,132 @@ let get_pool_updates_in_json ~__context ~hosts =
   | e ->
     error "getting updates for pool failed: %s" (ExnHelper.string_of_exn e);
     raise Api_errors.(Server_error (get_updates_failed, []))
+
+let apply ~__context ~host =
+  (* This function runs on slave host *)
+  with_local_repository ~__context (fun () ->
+      let params =
+          [
+            "-y"; "--disablerepo=*";
+            Printf.sprintf "--enablerepo=%s" !Xapi_globs.local_repo_name;
+            "upgrade"
+          ]
+      in
+      try ignore (Helpers.call_script !Xapi_globs.yum_cmd params)
+      with e ->
+        let ref = Ref.string_of host in
+        error "Failed to apply updates on host ref='%s': %s" ref (ExnHelper.string_of_exn e);
+        raise Api_errors.(Server_error (apply_updates_failed, [ref])))
+
+let set_of_list l =
+  l
+  |> List.map string_of_guidance
+  |> GuidanceStrSet.of_list
+
+let set1 = [EvacuateHost; RestartToolstack]
+
+let set2 = [RestartDeviceModel; RestartToolstack]
+
+let set3 = [RestartDeviceModel; EvacuateHost]
+
+let set4 = [EvacuateHost; RestartToolstack; RestartDeviceModel]
+
+let eq l s = GuidanceStrSet.equal (set_of_list l) (set_of_list s)
+
+let assert_valid_guidances = function
+  | [RebootHost]
+  | [EvacuateHost]
+  | [RestartToolstack]
+  | [RestartDeviceModel] -> ()
+  | l when eq l set1 -> ()
+  | l when eq l set2 -> ()
+  | l when eq l set3 -> ()
+  | l when eq l set4 -> ()
+  | l ->
+    let msg =
+      Printf.sprintf "Found wrong guidance(s) before applying updates: %s"
+        (String.concat ";" (List.map string_of_guidance l))
+    in
+    raise Api_errors.(Server_error (internal_error, [msg]))
+
+let apply_immediate_guidances ~__context ~host guidances =
+  try
+    let num_of_host = List.length (Db.Host.get_all ~__context) in
+    let open Client in
+    Helpers.call_api_functions ~__context (fun rpc session_id ->
+        match guidances with
+        | [RebootHost] ->
+          Client.Host.reboot ~rpc ~session_id ~host
```

Comment:
What if host = localhost (master) for reboot and restart_agent? Are we certain that there is nothing left to do after this function call? For example, we may miss the "cleanup" part of any with_ functions that we are nested inside, such as with_host_in_updating.

Reply:
I think they have 10-sec delay for XAPI call to complete remaining works.
But I defer the `Host.reboot` as reboot conflicts with apply_updates.

Reply:
I think `restart_agent` at least just executes `xe-toolstack-restart` immediately.

Reply:
Is this still true?
https://github.com/xapi-project/xen-api/blob/master/ocaml/idl/datamodel_host.ml#L438

Reply:
I don't think so, looking at `Xapi_host.restart_agent`.